Micrometer Observation

marcingrzejszczak opened the pull request:
fixes https://issues.couchbase.com/browse/JCBC-2046

This results in automatic creation of spans and metrics (and whatever else you configure through `ObservationRegistry`). Below you can find an example of the setup for Metrics with Micrometer Core and Tracing with Micrometer Tracing.

```
15:03:31 INFO  [io.micrometer.tracing.test.SampleTestRunner:183] Gathered the following metrics
	Meter with name <db.couchbase.operations.active> and type <LONG_TASK_TIMER> has the following measurements 
		<[Measurement{statistic='ACTIVE_TASKS', value=0.0}, Measurement{statistic='DURATION', value=0.0}]> 
		and has the following tags <[]>
	Meter with name <db.couchbase.operations> and type <TIMER> has the following measurements 
		<[Measurement{statistic='COUNT', value=1.0}, Measurement{statistic='TOTAL_TIME', value=0.002471388}, Measurement{statistic='MAX', value=0.002471388}]> 
		and has the following tags <[tag(db.system=couchbase), tag(error=none), tag(net.transport=IP.TCP)]>
	Meter with name <zipkin_brave.active> and type <LONG_TASK_TIMER> has the following measurements 
		<[Measurement{statistic='ACTIVE_TASKS', value=1.0}, Measurement{statistic='DURATION', value=0.178993099}]> 
		and has the following tags <[]>
	Meter with name <db.couchbase.operations> and type <TIMER> has the following measurements 
		<[Measurement{statistic='COUNT', value=1.0}, Measurement{statistic='TOTAL_TIME', value=0.003121377}, Measurement{statistic='MAX', value=0.003121377}]> 
		and has the following tags <[tag(db.couchbase.service=kv), tag(db.operation=get), tag(db.system=couchbase), tag(error=CompletionException)]>
```

![image](https://user-images.githubusercontent.com/3297437/233379142-ab005632-a31c-4278-9a6c-6dbfa8276f60.png)

programmatix commented on tracing-observation/src/main/java/com/couchbase/client/tracing/observation/CouchbaseSenderContext.java:
```diff
@@ -0,0 +1,41 @@
+ *
+ * @author Marcin Grzejszczak
+ */
+public class CouchbaseSenderContext extends SenderContext<Object> {
```
Could you provide more detail on what this is up to, as I couldn't find much online on a quick search for SenderContext. 

Is it responsible for sending traceparent from https://www.w3.org/TR/trace-context/?

programmatix replied:
Ok thanks.  That's pretty much what I'd guessed, just wanted to confirm.  (And TIL about the b3 form of trace propagation)

programmatix reviewed (changes requested):
First of all, I want to apologise for how long it's taken to get to this review, and to thank you for this contribution.

I think overall this looks like a good & well-implemented change, and reasonably straightforward.  There's a bit of feedback below but it's all minor.

One more major reservation I have is that I'm not sure how well it fits our model.  We've split out metrics and tracing into separate interfaces - Meter+ValueRecorder and RequestTracer+RequestSpan respectively.
But if I understand Micrometer Observation correctly, it bridges both of those concepts.
We seem to end up with the user creating a RequestTracer, and maybe assuming that's only to be outputting spans, when in reality it's also now outputting metrics.
It also raises questions of what if, for some reason, a user wants to have their spans going to Micrometer and their metrics to a different backend, like OpenTelemetry.

I'm not sure how best to resolve this...

marcingrzejszczak commented:
So Micrometer Observation doesn't care which metrics library is being used. You can use micrometer observation with an otel metrics handler.

marcingrzejszczak commented on README.md:
```diff
@@ -21,6 +21,7 @@ This repository contains the following projects:
  - `java-client`: the Java language binding
  - `scala-client`: the Scala language binding
  - `kotlin-client`: the Kotlin language binding
+ - `tracing-observation`: module to integrate with [Micrometer Observation](https://micrometer.io/docs/observation)
```
Actually we don't know what that will do because it all depends on the configuration of the handler mechanism. I added the tracing prefix because i'm using the tracing api.

marcingrzejszczak commented:
Sorry for a short reply yesterday, I've answered your comments from my phone :)

> First of all, I want to apologise for how long it's taken to get to this review, and to thank you for this contribution.

You have nothing to apologize for.

> I think overall this looks like a good & well-implemented change, and reasonably straightforward. There's a bit of feedback below but it's all minor.

Thanks!

> One more major reservation I have is that I'm not sure how well it fits our model. We've split out metrics and tracing into separate interfaces - Meter+ValueRecorder and RequestTracer+RequestSpan respectively.

Conceptually the tracer api made more sense.

> But if I understand Micrometer Observation correctly, it bridges both of those concepts.

It can bridge both, or none or other concepts. It's up to the user to configure the handlers.

> We seem to end up with the user creating a RequestTracer, and maybe assuming that's only to be outputting spans, when in reality it's also now outputting metrics.

Yeah, they might not even be outputting spans at all.

> It also raises questions of what if, for some reason, a user wants to have their spans going to Micrometer and their metrics to a different backend, like OpenTelemetry.

No problem, they can configure the handlers to use whatever they want for metrics, traces, logging, etc.

> I'm not sure how best to resolve this...

From my perspective Micrometer Observation abstracts all pillars of observability in such a way that you instrument once but via the handler configuration you can define the features you want to have for that instrumentation.

BTW I've rebased against master and hopefully fixed all the conflicts accordingly.

programmatix commented on tracing-observation/src/integrationTest/java/com/couchbase/client/tracing/observation/ObservationTracingIntegrationTest.java:
```diff
@@ -0,0 +1,156 @@
+    TestNodeConfig config = config().firstNodeWith(Services.KV).get();
+
+    cluster = Cluster.connect(
+      String.format("couchbase://%s:%d", config.hostname(), config.ports().get(Services.KV)),
```
Ah, that's fair enough.  It's not a huge deal, it's just unusual to see.  It can be left

programmatix commented:
Thank you @marcingrzejszczak for taking the time to reply to all my questions.  I want to stress that overall I do like this change, very much appreciate the contribution, and want to get it into the codebase in some form.  The only things giving me pause are:

1. The issue described above - that we've separated out metrics and tracing into separate interfaces, but that Micrometer Observation doesn't fit that model well since it can be one, or either, or none of those. 
I agree with you that putting Observation into the tracing interface is definitely the better fit out of tracing & metrics.
I suppose the only real other approach is to force a lot of complexity onto the user.  They would have to create two `ObservationRegistry`s, passing one each to our tracing and metrics interface.  But that doesn't feel user friendly, and it doesn't feel in the spirit of Micrometer Observation.
So, I guess I'm coming round to the inevitability of this approach.

2. Adding support for a Micrometer-only feature (high vs low cardinality attributes) into the generic RequestSpan interface.
Partly as it feels a little breaky, in that users may be setting attributes that are either high or low, but have now been retroactively decided to be high.
Partly as it's a feature that only Micrometer supports, AFAIK.  If we supported every feature of every tracing/metrics backend then the interface would be vast, so we try to keep it to the highest-impact features.
And partly as it's impacting on the RequestSpan interface shared by all SDKs, as generally we want to have those specced interfaces identical between SDKs.  Though we could well argue this is a special case, since it's purely to support JVM-only Micrometer.

I understand what high vs low cardinality attributes are but I couldn't find any info in the Micrometer docs on how these are used, and what the impact is of using them.  The popular metrics backends I'm most familiar with, such as Prometheus, don't have any support for the concept AFAIK.  Does it come down simply to this?
```
Observation.start("my.operation", registry)
    .lowCardinalityKeyValue("this.tag", "will end up as a meter tag and a span tag")
    .highCardinalityKeyValue("but.this.tag", "will end up as a span tag only");
```

marcingrzejszczak commented:
> Thank you @marcingrzejszczak for taking the time to reply to all my questions. I want to stress that overall I do like this change, very much appreciate the contribution, and want to get it into the codebase in some form. The only things giving me pause are:

My pleasure :)

> The issue described above - that we've separated out metrics and tracing into separate interfaces, but that Micrometer Observation doesn't fit that model well since it can be one, or either, or none of those.
> I agree with you that putting Observation into the tracing interface is definitely the better fit out of tracing & metrics.
> I suppose the only real other approach is to force a lot of complexity onto the user. They would have to create two ObservationRegistrys, passing one each to our tracing and metrics interface. But that doesn't feel user friendly, and it doesn't feel in the spirit of Micrometer Observation.
> So, I guess I'm coming round to the inevitability of this approach.

Yeah that wouldn't make sense and would be counter-intuitive really :shrug: 

> I understand what high vs low cardinality attributes are but I couldn't find any info in the Micrometer docs on how these are used, and what the impact is of using them. The popular metrics backends I'm most familiar with, such as Prometheus, don't have any support for the concept AFAIK. Does it come down simply to this?

Yes, it boils down to that but not only to that. If you put high cardinality data to Prometheus [you can blow up Prometheus.](https://prometheus.io/docs/practices/instrumentation/#do-not-overuse-labels). Check also [this article](https://grafana.com/blog/2022/02/15/what-are-cardinality-spikes-and-why-do-they-matter/). So Prometheus doesn't really expect you to put high cardinality data by design, that's why it's not in the API. Micrometer Observation is the only API that I'm aware of that allows you to work on both metrics and tracing level at the same time in which case you have to start working on the cardinality level too.

programmatix commented:
Yea, I'm familiar with the cardinality issue with Prometheus.  

> Micrometer Observation is the only API that I'm aware of that allows you to work on both metrics and tracing level at the same time in which case you have to start working on the cardinality level too.

Right, but it's another example of the friction I mentioned earlier - things are being added into the shared tracing API (RequestSpan), when really the use-case is for metrics.

I'm going to seek out another opinion here at Couchbase, and see if we can find a way forward.  Possibly I'm just overthinking this!

marcingrzejszczak commented:
> Right, but it's another example of the friction I mentioned earlier - things are being added into the shared tracing API (RequestSpan), when really the use-case is for metrics.

You're absolutely right!

> I'm going to seek out another opinion here at Couchbase, and see if we can find a way forward. Possibly I'm just overthinking this!

Sure, it's never enough to get second opinions :)

dnault commented on tracing-observation/src/main/java/com/couchbase/client/tracing/observation/ObservationRequestSpan.java:
```diff
@@ -0,0 +1,118 @@
+/**
+ * Wraps a Micrometer Observation, ready to be passed in into options for each operation into the SDK as a parent.
+ *
+ * @author Marcin Grzejszczak
```
Hi! I'm one of Graham's colleagues. As a team, we've been gradually removing Javadoc author tags from our codebase, since the version control history provides a more useful record of authorship. That said, I'll echo Graham's sentiment that your contribution is appreciated, and we'll be sure to credit you in the release notes unless you'd rather we didn't. :-)

marcingrzejszczak replied:
Lol😀 of course I'll remove the author tag, that's not a problem - we do it in micrometer and spring that's why I did here 🤷‍♂️

marcingrzejszczak replied:
Fixed!

programmatix commented:
@marcingrzejszczak David and I discussed this at length yesterday, and have found a path through the concerns raised above.  

We've decided that having high-cardinality and low-cardinality attributes in the user-facing RequestTracer interface - the primary blocker - is ok, as cardinality is a generic concept that could apply to tracing as well.  It's true that no current tracing backend differentiates between high & low cardinality - but a future one could.

However, it is possible for users to use the RequestTracer interface to create spans in their own app, and we do want to avoid confusing them with needing to understand the difference between the two - and then further explaining that it's a concept that in practice is only used by Micrometer Observation, and only for metrics.

So we have two requests, and the first is just to apply the @Stability.Internal annotation on the new lowCardinalityAttribute() methods in RequestTracer.  The user then has a clue that they can safely ignore them.

Request two is over the module naming.  "tracing-observation" just looks too generic to us, and we'd prefer "tracing-micrometer-observation".  It better fits our existing modules such as "tracing-opentelemetry" and "metrics-micrometer".  Generally we have the name of the product in there.

If you're happy to make those changes, we'll happily take this PR.

(Though I do have a feeling we have to have it in Gerrit and can't take a PR directly... I'm researching that now.)

marcingrzejszczak commented:
Hey! Thanks for the review. All changes got incorporated :)

programmatix commented:
Perfect, thank you!  Then just for internal confirmation on whether we've got a path to accept a PR directly, or whether it has to go through Gerrit.  Sorry, I should have raised this earlier.  Just want to make sure you get the credit for all this work rather than us merging it for you (though we can do that too if you prefer).

marcingrzejszczak commented:
I mean whatever you prefer :) I would prefer my name to be present somewhere but for the sake of support (so that someone can mention me if something is not working the way it should).

programmatix commented:
Apologies @marcingrzejszczak but I've checked and the review does have to go through Gerrit rather than Github.  I can do that for you but it would go under my name rather than yours, and as you say, it would be much better for it to be associated with yours.  Would you mind?  

https://github.com/couchbase/couchbase-jvm-clients/blob/master/CONTRIBUTING.md details how to signup on Gerrit and sign the Contributor License Agreement (CLA).

https://github.com/couchbase/couchbase-java-client/blob/master/CONTRIBUTING.md details how to push a change to Gerrit.

Just one tweak - the site is now secured so need to use `wget --no-check-certificate http://review.couchbase.com/tools/hooks/commit-msg`.

marcingrzejszczak commented:
Hey I think I've prepared my account and signed everything I should and still 

> You can start by submitting a pull request to the GitHub repo. A bot will comment on your PR and prompt you to register for a Gerrit account and sign a Contributor License Agreement (CLA).

there is no bot to comment on the PR :shrug:

marcingrzejszczak commented:
I think I've managed to do it! https://review.couchbase.org/c/couchbase-jvm-clients/+/196522 . IMO the readme should provide a link on how to clone gerrits repository or some sort of a hint on how to go from github to gerrit. Anyways, I think it works now :)
